File: 03-06-2026/schema_validation4.js
function isValidSchema(obj) {
  const ROLES = ["user", "creator", "moderator", "staff", "admin"];
  const SCHEMA = [
    ["username", "string"],
    ["posts", "number"],
    ["verified", "boolean"],
    ["role", "string"],
    ["supporter", "boolean", true],
  ];
  const isValidProp = (obj, key, expected_type, optional = false) =>
    optional
      ? obj.hasOwnProperty(key)
        ? typeof obj[key] === expected_type
        : true
      : obj.hasOwnProperty(key) && typeof obj[key] === expected_type;

  const hasValidRole = ROLES.includes(obj.role);
  const requiredProps = SCHEMA.map(([key, type, optional]) =>
    isValidProp(obj, key, type, optional),
  );

  return requiredProps.concat([hasValidRole]).every((elm) => elm);
}

console.log(
  isValidSchema({
    username: "vivian",
    posts: 1,
    verified: false,
    role: "user",
    supporter: true,
  }),
); // true.
console.log(
  isValidSchema({
    username: "rudolph",
    posts: 15,
    verified: true,
    role: "creator",
  }),
); // true.
console.log(
  isValidSchema({
    username: "hernandez",
    posts: 35,
    verified: true,
    role: "moderator",
    supporter: false,
    followers: 55,
  }),
); // true.
console.log(
  isValidSchema({
    username: "julia",
    posts: 50,
    verified: true,
    role: "admin",
    supporter: "true",
  }),
); // false.
console.log(
  isValidSchema({
    username: "bernard",
    posts: 0,
    verified: true,
    role: "friend",
    supporter: true,
  }),
); // false.
console.log(
  isValidSchema({
    username: "felix",
    posts: 40,
    verified: "yes",
    role: "staff",
    supporter: false,
  }),
); // false.
console.log(
  isValidSchema({
    username: "jimmy",
    posts: true,
    verified: false,
    role: "creator",
    supporter: true,
  }),
); // false.
console.log(
  isValidSchema({
    username: true,
    posts: 30,
    verified: true,
    role: "moderator",
    supporter: false,
  }),
); // false.

File: 03-06-2026/schema_validation4.py
def is_valid_schema(obj):
    ROLES = ["user", "creator", "moderator", "staff", "admin"];
    SCHEMA = [
        ["username", str, False],
        ["posts", int, False],
        ["verified", bool, False],
        ["role", str, False],
        ["supporter", bool, True]
    ]

    if obj["role"] not in ROLES:
        return False
    
    for key, expected_type, optional in SCHEMA:
        if not is_valid(obj, key, expected_type, optional=optional):
            return False

    return True

def is_valid(obj, key, expected_type, optional):
    if optional:
        if (key in obj and isinstance(obj[key], expected_type)) or key not in obj:
            return True
    elif key in obj:
        if expected_type == int and isinstance(obj[key], bool):
            return False 
        elif expected_type == bool and isinstance(obj[key], int) and not isinstance(obj[key], bool):
            return False
        elif isinstance(obj[key], expected_type):
            return True
    
    return False

print(is_valid_schema({"username": "vivian", "posts": 1, "verified": False, "role": "user", "supporter": True})) # True.
print(is_valid_schema({"username": "rudolph", "posts": 15, "verified": True, "role": "creator"})) # True.
print(is_valid_schema({"username": "hernandez", "posts": 35, "verified": True, "role": "moderator", "supporter": False, "followers": 55})) # True.
print(is_valid_schema({"username": "julia", "posts": 50, "verified": True, "role": "admin", "supporter": "true"})) # False.
print(is_valid_schema({"username": "bernard", "posts": 0, "verified": True, "role": "friend", "supporter": True})) # False.
print(is_valid_schema({"username": "felix", "posts": 40, "verified": "yes", "role": "staff", "supporter": False})) # False.
print(is_valid_schema({"username": "jimmy", "posts": True, "verified": False, "role": "creator", "supporter": True})) # False.
print(is_valid_schema({"username": True, "posts": 30, "verified": True, "role": "moderator", "supporter": False})) # False.